### PR TITLE
[BugFix] Fix NoopResetEnv behavior when trials exceeded.

### PR DESCRIPTION
**Pull request by skandermoalla**
## Description

When the `_MAX_NOOPS_TRIALS` were exceeded the transform was calling a final `rand_step()` without a `step_mdp()`.
This PR add the `step_mdp()` call and 
- Improves the error message.
- Adds tests to test for the edge case behaviors of the transform.
- Fixes the documentation of the tranform.

Some things could be discussed. I left a TODO in front, which we will remove before merging.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/envs/transforms/transforms.py [hunk 1]
```diff
@@ -3167,23 +3168,40 @@ def reset(self, tensordict: TensorDictBase) -> TensorDictBase:
...
             trial += 1
             if trial > _MAX_NOOPS_TRIALS:
+                trials_exceeded_message = (
+                    "Parent env was repeatedly done or truncated"
+                    " before the sampled number of noops could be applied. "
+                )
                 tensordict = parent.rand_step(tensordict)
-                if tensordict.get(("next", done_key)):
+                tensordict = step_mdp(tensordict, exclude_done=False)
+                if tensordict.get(done_key) or tensordict.get(
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:3187]: As of now this works only if done is a scalar. Should we contemplate making this transform compatible with batched envs? Not sure how it would look like tbh

File: torchrl/envs/transforms/transforms.py [hunk 2]
```diff
@@ -3167,23 +3168,40 @@ def reset(self, tensordict: TensorDictBase) -> TensorDictBase:
...
+                    )
+                else:
+                    warnings.warn(
+                        trials_exceeded_message + "A single noop step was applied."
                     )
+                    # TODO: up to discussion. Could also be reset instead of doing a single noop.
                 break
 
         if tensordict.get(done_key):
+            # TODO: This will never be reached. It should be removed.
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:3202]: I agree let's ditch it

**Review by vmoens — APPROVED** [review]
LGTM thanks for this. IMO we don't want to reset if we reach the max number of trials bc it could lead to a silent bug. WDYT?

**Comment by skandermoalla** [comment]
Yes. I also think a single reset is arbitrary and would lead to issues. That was the original behavior, I just fixed a bug in it.

I will refactor the transform so that it becomes simpler to parse and has less edge cases.

File: torchrl/envs/transforms/transforms.py (hunk 1 above)
Comment (skandermoalla) [reply torchrl/envs/transforms/transforms.py:3187]: The transform assumes a single env as a parent at the moment:

\- \[Code\](https://github.com/pytorch/rl/blob/7ae614043b67bfccf849f661491587339e3262a2/torchrl/envs/transforms/transforms.py#L3148).

\- \[test\](https://github.com/pytorch/rl/blob/7ae614043b67bfccf849f661491587339e3262a2/test/test_transforms.py#L3072) 

This change will require a refactor of the whole transform, so I would keep it out of the scopre of this PR.

**Comment by skandermoalla** [comment]
With this, the transform raises an error whenver its consistently not possible to ally the noops (1). We could also just raise a warning and not do any noop at reset (2).

Both options are valid in my opinion, and maybe the first one is better as it will nudge to user to be more careful when using the transform. E.g.

With (1) the user cannot get away by setting a high `noops` and \`random=True\` that may fail occasionally. With (2) they could get away with it at the env will just not receive `noops` occasinally.
